plugins/microsoft-graph-authz/src/main/java/org/elasticsearch/xpack/security/authz/microsoft/MicrosoftGraphAuthzRealm.java

```diff
@@ -43,8 +126,112 @@ public void authenticate(AuthenticationToken token, ActionListener<Authenticatio
     }
 
     @Override
-    public void lookupUser(String username, ActionListener<User> listener) {
-        logger.info("Microsoft Graph Authz not yet implemented, returning empty roles for [{}]", username);
-        listener.onResponse(new User(username));
+    public void lookupUser(String principal, ActionListener<User> listener) {
+        if (MICROSOFT_GRAPH_FEATURE.check(licenseState) == false) {
+            listener.onFailure(LicenseUtils.newComplianceException(MICROSOFT_GRAPH_FEATURE.getName()));
+            return;
+        }
+
+        final var runnable = new CancellableRunnable<>(listener, ex -> null, () -> doLookupUser(principal, listener), logger);
+        threadPool.generic().execute(runnable);
+        threadPool.schedule(runnable::maybeTimeout, executionTimeout, EsExecutors.DIRECT_EXECUTOR_SERVICE);
+    }
+
+    private void doLookupUser(String principal, ActionListener<User> listener) {
+        try {
+            final var userProperties = fetchUserProperties(client, principal);
+            final var groups = fetchGroupMembership(client, principal);
```

Comment:
Not to wanting to create more work, but ...

It would be awesome if the cancellation could be checked in between these API calls, rather than just at the start of the runnable.
We could do that by replacing `Runnable` with `Consumer<AtomicBoolean>` where the boolean is `shouldCancel`.

That's a big enough change that we shouldn't try and squeeze it into this PR, but I think it would be a nice improvement at some point.

Reply:
Created a jira for this improvement: ES-12057